File: app/src/main/java/com/xeniac/warrantyroster/mainactivity/addwarrantyfragment/AddWarrantyFragment.java
package com.xeniac.warrantyroster.mainactivity.addwarrantyfragment;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.Editable;
import android.text.InputType;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.request.RequestHeaders;
import com.google.android.material.datepicker.MaterialDatePicker;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.xeniac.warrantyroster.Constants;
import com.xeniac.warrantyroster.NetworkHelper;
import com.xeniac.warrantyroster.R;
import com.xeniac.warrantyroster.database.CategoryDataModel;
import com.xeniac.warrantyroster.database.WarrantyRosterDatabase;
import com.xeniac.warrantyroster.databinding.FragmentAddWarrantyBinding;
import com.xeniac.warrantyroster.mainactivity.MainActivity;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.List;

public class AddWarrantyFragment extends Fragment {

    private FragmentAddWarrantyBinding addWarrantyBinding;
    private View view;
    private Activity activity;
    private Context context;
    private NavController navController;
    private WarrantyRosterDatabase database;

    private final DecimalFormat decimalFormat = new DecimalFormat("00");
    private CategoryDataModel selectedCategory;
    private Calendar startingCalendar, expiryCalendar;
    private String startingDateInput, expiryDateInput;

    public AddWarrantyFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        addWarrantyBinding = FragmentAddWarrantyBinding.inflate(inflater, container, false);
        view = addWarrantyBinding.getRoot();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        WarrantyRosterDatabase.destroyInstance();
        addWarrantyBinding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        activity = getActivity();
        context = getContext();
        navController = Navigation.findNavController(view);
        database = WarrantyRosterDatabase.getInstance(context);
        ((MainActivity) context).hideNavBar();

        textInputsBackgroundColor();
        textInputsStrokeColor();
        categoryDropDownSelection();
        startingDatePicker();
        expiryDatePicker();
        returnToMainActivity();
        addWarrantyOnClick();
    }

    @Override
    public void onResume() {
        super.onResume();
        categoryDropDown();
    }

    private void textInputsBackgroundColor() {
        addWarrantyBinding.tiAddWarrantyEditTitle.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                addWarrantyBinding.tiAddWarrantyLayoutTitle.setBoxBackgroundColorResource(R.color.background);
            } else {
                addWarrantyBinding.tiAddWarrantyLayoutTitle.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });

        addWarrantyBinding.tiAddWarrantyDdCategory.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                addWarrantyBinding.tiAddWarrantyLayoutCategory.setBoxBackgroundColorResource(R.color.background);
            } else {
                addWarrantyBinding.tiAddWarrantyLayoutCategory.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });

        addWarrantyBinding.tiAddWarrantyEditBrand.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                addWarrantyBinding.tiAddWarrantyLayoutBrand.setBoxBackgroundColorResource(R.color.background);
            } else {
                addWarrantyBinding.tiAddWarrantyLayoutBrand.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });

        addWarrantyBinding.tiAddWarrantyEditModel.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                addWarrantyBinding.tiAddWarrantyLayoutModel.setBoxBackgroundColorResource(R.color.background);
            } else {
                addWarrantyBinding.tiAddWarrantyLayoutModel.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });

        addWarrantyBinding.tiAddWarrantyEditSerial.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                addWarrantyBinding.tiAddWarrantyLayoutSerial.setBoxBackgroundColorResource(R.color.background);
            } else {
                addWarrantyBinding.tiAddWarrantyLayoutSerial.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });

        addWarrantyBinding.tiAddWarrantyEditDateStarting.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                addWarrantyBinding.tiAddWarrantyLayoutDateStarting.setBoxBackgroundColorResource(R.color.background);
            } else {
                addWarrantyBinding.tiAddWarrantyLayoutDateStarting.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });

        addWarrantyBinding.tiAddWarrantyEditDateExpiry.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                addWarrantyBinding.tiAddWarrantyLayoutDateExpiry.setBoxBackgroundColorResource(R.color.background);
            } else {
                addWarrantyBinding.tiAddWarrantyLayoutDateExpiry.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });

        addWarrantyBinding.tiAddWarrantyEditDescription.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                addWarrantyBinding.tiAddWarrantyLayoutDescription.setBoxBackgroundColorResource(R.color.background);
            } else {
                addWarrantyBinding.tiAddWarrantyLayoutDescription.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });
    }

    private void textInputsStrokeColor() {
        addWarrantyBinding.tiAddWarrantyEditDateStarting.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                hideDateError();
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });

        addWarrantyBinding.tiAddWarrantyEditDateExpiry.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                hideDateError();
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });
    }

    private void categoryDropDown() {
        List<String> titlesList = new ArrayList<>();
        for (Integer categoryTitleId : database.categoryDAO().getAllCategoryTitles()) {
            titlesList.add(context.getResources().getString(categoryTitleId));
        }

        addWarrantyBinding.tiAddWarrantyDdCategory.setAdapter(new ArrayAdapter<>(context, R.layout.dropdown_category, titlesList));
    }

    private void categoryDropDownSelection() {
        addWarrantyBinding.tiAddWarrantyDdCategory.setOnItemClickListener((adapterView, view, index, l) -> {
            selectedCategory = database.categoryDAO().getCategoryById(String.valueOf(index + 1));
            if (selectedCategory != null) {
                addWarrantyBinding.ivAddWarrantyIconCategory.setImageDrawable(ResourcesCompat.getDrawable(
                        getResources(), selectedCategory.getIcon(), context.getTheme()));
            }
        });
    }

    private void startingDatePicker() {
        addWarrantyBinding.tiAddWarrantyEditDateStarting.setInputType(InputType.TYPE_NULL);
        addWarrantyBinding.tiAddWarrantyEditDateStarting.setKeyListener(null);

        addWarrantyBinding.tiAddWarrantyEditDateStarting.setOnFocusChangeListener((view, isFocused) -> {
            if (isFocused) {
                openStartingDatePicker();
            }
        });
    }

    private void expiryDatePicker() {
        addWarrantyBinding.tiAddWarrantyEditDateExpiry.setInputType(InputType.TYPE_NULL);
        addWarrantyBinding.tiAddWarrantyEditDateExpiry.setKeyListener(null);

        addWarrantyBinding.tiAddWarrantyEditDateExpiry.setOnFocusChangeListener((view, isFocused) -> {
            if (isFocused) {
                openExpiryDatePicker();
            }
        });
    }

    private void openStartingDatePicker() {
        MaterialDatePicker<Long> startingDP = MaterialDatePicker.Builder.datePicker()
                .setTitleText(getResources().getString(R.string.add_warranty_title_date_picker_starting))
                .setInputMode(MaterialDatePicker.INPUT_MODE_CALENDAR)
                .setSelection(MaterialDatePicker.todayInUtcMilliseconds())
                .build();
        startingDP.show(getParentFragmentManager(), Constants.FRAGMENT_TAG_ADD_CALENDAR_STARTING);

        startingDP.addOnPositiveButtonClickListener(selection -> {
            startingCalendar = Calendar.getInstance();
            startingCalendar.setTimeInMillis(selection);
            startingDateInput = String.format("%s-%s-%s",
                    startingCalendar.get(Calendar.YEAR),
                    decimalFormat.format((startingCalendar.get(Calendar.MONTH)) + 1),
                    decimalFormat.format(startingCalendar.get(Calendar.DAY_OF_MONTH)));

            String startingDateInput = String.format("%s/%s/%s",
                    decimalFormat.format(startingCalendar.get(Calendar.DAY_OF_MONTH)),
                    decimalFormat.format((startingCalendar.get(Calendar.MONTH)) + 1),
                    startingCalendar.get(Calendar.YEAR));

            addWarrantyBinding.tiAddWarrantyEditDateStarting.setText(startingDateInput);
            addWarrantyBinding.tiAddWarrantyEditDateStarting.clearFocus();
        });

        startingDP.addOnDismissListener(dialogInterface ->
                addWarrantyBinding.tiAddWarrantyEditDateStarting.clearFocus());
    }

    private void openExpiryDatePicker() {
        MaterialDatePicker<Long> expiryDP = MaterialDatePicker.Builder.datePicker()
                .setTitleText(getResources().getString(R.string.add_warranty_title_date_picker_expiry))
                .setInputMode(MaterialDatePicker.INPUT_MODE_CALENDAR)
                .build();
        expiryDP.show(getParentFragmentManager(), Constants.FRAGMENT_TAG_ADD_CALENDAR_EXPIRY);

        expiryDP.addOnPositiveButtonClickListener(selection -> {
            expiryCalendar = Calendar.getInstance();
            expiryCalendar.setTimeInMillis(selection);
            expiryDateInput = String.format("%s-%s-%s",
                    expiryCalendar.get(Calendar.YEAR),
                    decimalFormat.format((expiryCalendar.get(Calendar.MONTH)) + 1),
                    decimalFormat.format(expiryCalendar.get(Calendar.DAY_OF_MONTH)));

            String expiryDateInput = String.format("%s/%s/%s",
                    decimalFormat.format(expiryCalendar.get(Calendar.DAY_OF_MONTH)),
                    decimalFormat.format((expiryCalendar.get(Calendar.MONTH)) + 1),
                    expiryCalendar.get(Calendar.YEAR));

            addWarrantyBinding.tiAddWarrantyEditDateExpiry.setText(expiryDateInput);
            addWarrantyBinding.tiAddWarrantyEditDateExpiry.clearFocus();
        });

        expiryDP.addOnDismissListener(dialogInterface ->
                addWarrantyBinding.tiAddWarrantyEditDateExpiry.clearFocus());
    }

    private void returnToMainActivity() {
        addWarrantyBinding.toolbarAddWarranty.setNavigationOnClickListener(view ->
                activity.onBackPressed());
    }

    private void addWarrantyOnClick() {
        addWarrantyBinding.toolbarAddWarranty.getMenu().getItem(0).setOnMenuItemClickListener(menuItem -> {
            addWarranty();
            return false;
        });
    }

    private void addWarranty() {
        InputMethodManager inputMethodManager = (InputMethodManager)
                context.getSystemService(Context.INPUT_METHOD_SERVICE);
        inputMethodManager.hideSoftInputFromWindow(view.getApplicationWindowToken(), 0);

        if (NetworkHelper.hasNetworkAccess(context)) {
            getWarrantyInput();
        } else {
            hideLoadingAnimation();
            Snackbar.make(view, context.getResources().getString(R.string.network_error_connection),
                    BaseTransientBottomBar.LENGTH_INDEFINITE)
                    .setAction(context.getResources().getString(R.string.network_error_retry), v -> addWarranty()).show();
        }
    }

    private void getWarrantyInput() {
        if (TextUtils.isEmpty(addWarrantyBinding.tiAddWarrantyEditTitle.getText())) {
            addWarrantyBinding.tiAddWarrantyLayoutTitle.requestFocus();
            addWarrantyBinding.tiAddWarrantyLayoutTitle.setBoxStrokeColor(context.getResources().getColor(R.color.red));
        } else if (startingCalendar == null) {
            addWarrantyBinding.tiAddWarrantyLayoutDateStarting.requestFocus();
            addWarrantyBinding.tiAddWarrantyLayoutDateStarting.setBoxStrokeColor(context.getResources().getColor(R.color.red));
        } else if (expiryCalendar == null) {
            addWarrantyBinding.tiAddWarrantyLayoutDateExpiry.requestFocus();
            addWarrantyBinding.tiAddWarrantyLayoutDateExpiry.setBoxStrokeColor(context.getResources().getColor(R.color.red));
        } else if (!isStartingDateValid(startingCalendar, expiryCalendar)) {
            showDateError();
        } else {
            String title = addWarrantyBinding.tiAddWarrantyEditTitle.getText().toString().trim();
            String brand = null;
            String model = null;
            String serialNumber = null;
            String description = null;
            String categoryId = "10";

            if (!TextUtils.isEmpty(addWarrantyBinding.tiAddWarrantyEditBrand.getText())) {
                brand = addWarrantyBinding.tiAddWarrantyEditBrand.getText().toString().trim();
            }
            if (!TextUtils.isEmpty(addWarrantyBinding.tiAddWarrantyEditModel.getText())) {
                model = addWarrantyBinding.tiAddWarrantyEditModel.getText().toString().trim();
            }
            if (!TextUtils.isEmpty(addWarrantyBinding.tiAddWarrantyEditSerial.getText())) {
                serialNumber = addWarrantyBinding.tiAddWarrantyEditSerial.getText().toString().trim();
            }
            if (!TextUtils.isEmpty(addWarrantyBinding.tiAddWarrantyEditDescription.getText())) {
                description = addWarrantyBinding.tiAddWarrantyEditDescription.getText().toString().trim();
            }
            if (selectedCategory != null) {
                categoryId = selectedCategory.getId();
            }

            AddWarrantyInput warrantyInput = AddWarrantyInput.builder()
                    .title(title)
                    .brand(brand)
                    .model(model)
                    .serial_number(serialNumber)
                    .starting_date(startingDateInput)
                    .expiry_date(expiryDateInput)
                    .description(description)
                    .category_id(categoryId)
                    .build();

            addWarrantyMutation(warrantyInput);
        }
    }

    private void addWarrantyMutation(AddWarrantyInput warrantyInput) {
        showLoadingAnimation();

        SharedPreferences loginPrefs = context.getSharedPreferences(Constants.PREFERENCE_LOGIN, Context.MODE_PRIVATE);
        String userToken = loginPrefs.getString(Constants.PREFERENCE_USER_TOKEN_KEY, null);

        ApolloClient apolloClient = ApolloClient.builder()
                .serverUrl(Constants.URL_GRAPHQL)
                .build();

        apolloClient.mutate(new AddWarrantyMutation(warrantyInput))
                .toBuilder().requestHeaders(RequestHeaders.builder().addHeader("Authorization", "Bearer " + userToken).build())
                .build()
                .enqueue(new ApolloCall.Callback<AddWarrantyMutation.Data>() {
                    @Override
                    public void onResponse(@NonNull Response<AddWarrantyMutation.Data> response) {
                        if (addWarrantyBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();

                                if (!response.hasErrors()) {
                                    Log.i("addWarranty", "onResponse: " + response);
                                    navController.navigate(R.id.action_addWarrantyFragment_to_warrantiesFragment);
                                } else {
                                    Log.e("addWarranty", "onResponse Errors: " + response.getErrors());
                                    if (Arrays.toString(response.getErrors().get(0).getCustomAttributes().values().toArray()).contains("code=422")) {
                                        Snackbar.make(view, context.getResources().getString(R.string.add_warranty_error_date),
                                                BaseTransientBottomBar.LENGTH_LONG).show();
                                    } else {
                                        Snackbar.make(view, context.getResources().getString(R.string.network_error_response),
                                                BaseTransientBottomBar.LENGTH_INDEFINITE)
                                                .setAction(context.getResources().getString(R.string.network_error_retry), v -> addWarranty()).show();
                                    }
                                }
                            });
                        }
                    }

                    @Override
                    public void onFailure(@NonNull ApolloException e) {
                        Log.e("addWarranty", "onFailure: " + e.getMessage());
                        if (addWarrantyBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();
                                Snackbar.make(view, context.getResources().getString(R.string.network_error_failure),
                                        BaseTransientBottomBar.LENGTH_LONG).show();
                            });
                        }
                    }
                });
    }

    private void showLoadingAnimation() {
        addWarrantyBinding.tiAddWarrantyEditTitle.setEnabled(false);
        addWarrantyBinding.tiAddWarrantyDdCategory.setEnabled(false);
        addWarrantyBinding.tiAddWarrantyEditBrand.setEnabled(false);
        addWarrantyBinding.tiAddWarrantyEditModel.setEnabled(false);
        addWarrantyBinding.tiAddWarrantyEditSerial.setEnabled(false);
        addWarrantyBinding.tiAddWarrantyEditDateStarting.setEnabled(false);
        addWarrantyBinding.tiAddWarrantyEditDateExpiry.setEnabled(false);
        addWarrantyBinding.tiAddWarrantyEditDescription.setEnabled(false);
        addWarrantyBinding.cpiAddWarranty.setVisibility(View.VISIBLE);
    }

    private void hideLoadingAnimation() {
        addWarrantyBinding.tiAddWarrantyEditTitle.setEnabled(true);
        addWarrantyBinding.tiAddWarrantyDdCategory.setEnabled(true);
        addWarrantyBinding.tiAddWarrantyEditBrand.setEnabled(true);
        addWarrantyBinding.tiAddWarrantyEditModel.setEnabled(true);
        addWarrantyBinding.tiAddWarrantyEditSerial.setEnabled(true);
        addWarrantyBinding.tiAddWarrantyEditDateStarting.setEnabled(true);
        addWarrantyBinding.tiAddWarrantyEditDateExpiry.setEnabled(true);
        addWarrantyBinding.tiAddWarrantyEditDescription.setEnabled(true);
        addWarrantyBinding.cpiAddWarranty.setVisibility(View.GONE);
    }

    private void showDateError() {
        addWarrantyBinding.tiAddWarrantyLayoutDateStarting.requestFocus();
        addWarrantyBinding.tiAddWarrantyLayoutDateStarting.setBoxStrokeColor(context.getResources().getColor(R.color.red));
        addWarrantyBinding.tvAddWarrantyDateError.setVisibility(View.VISIBLE);
    }

    private void hideDateError() {
        addWarrantyBinding.tiAddWarrantyLayoutDateStarting.setBoxStrokeColor(context.getResources().getColor(R.color.blue));
        addWarrantyBinding.tvAddWarrantyDateError.setVisibility(View.GONE);
    }

    private boolean isStartingDateValid(Calendar startingCalendar, Calendar expiryCalendar) {
        return expiryCalendar.compareTo(startingCalendar) >= 0;
    }
}

File: app/src/main/java/com/xeniac/warrantyrostermanager/mainactivity/addwarrantyfragment/AddWarrantyFragment.kt
package com.xeniac.warrantyrostermanager.mainactivity.addwarrantyfragment

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.inputmethod.InputMethodManager
import android.widget.ArrayAdapter
import androidx.core.content.ContextCompat
import androidx.core.view.get
import androidx.fragment.app.Fragment
import androidx.navigation.NavController
import androidx.navigation.Navigation
import com.google.android.material.datepicker.MaterialDatePicker
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_INDEFINITE
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_LONG
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.ktx.auth
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import com.xeniac.warrantyrostermanager.Constants
import com.xeniac.warrantyrostermanager.NetworkHelper
import com.xeniac.warrantyrostermanager.R
import com.xeniac.warrantyrostermanager.database.WarrantyRosterDatabase
import com.xeniac.warrantyrostermanager.databinding.FragmentAddWarrantyBinding
import com.xeniac.warrantyrostermanager.mainactivity.MainActivity
import com.xeniac.warrantyrostermanager.model.Category
import com.xeniac.warrantyrostermanager.model.WarrantyInput
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import java.text.DecimalFormat
import java.util.*

class AddWarrantyFragment : Fragment(R.layout.fragment_add_warranty) {

    private var _binding: FragmentAddWarrantyBinding? = null
    private val binding get() = _binding!!
    private lateinit var navController: NavController
    private lateinit var database: WarrantyRosterDatabase
    private val warrantiesCollectionRef =
        Firebase.firestore.collection(Constants.COLLECTION_WARRANTIES)

    private val decimalFormat = DecimalFormat("00")
    private var selectedCategory: Category? = null
    private var startingCalendar: Calendar? = null
    private var expiryCalendar: Calendar? = null
    private lateinit var startingDateInput: String
    private lateinit var expiryDateInput: String

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentAddWarrantyBinding.bind(view)
        navController = Navigation.findNavController(view)
        database = WarrantyRosterDatabase.getInstance(requireContext())
        (requireContext() as MainActivity).hideNavBar()

        textInputsBackgroundColor()
        textInputsStrokeColor()
        categoryDropDownSelection()
        startingDatePicker()
        expiryDatePicker()
        returnToMainActivity()
        addWarrantyOnClick()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        WarrantyRosterDatabase.destroyInstance()
        _binding = null
    }

    override fun onResume() {
        super.onResume()
        categoryDropDown()
    }

    private fun textInputsBackgroundColor() {
        binding.tiAddWarrantyEditTitle.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiAddWarrantyLayoutTitle.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiAddWarrantyLayoutTitle.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }

        binding.tiAddWarrantyDdCategory.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiAddWarrantyLayoutCategory.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiAddWarrantyLayoutCategory.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }

        binding.tiAddWarrantyEditBrand.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiAddWarrantyLayoutBrand.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiAddWarrantyLayoutBrand.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }

        binding.tiAddWarrantyEditModel.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiAddWarrantyLayoutModel.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiAddWarrantyLayoutModel.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }

        binding.tiAddWarrantyEditSerial.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiAddWarrantyLayoutSerial.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiAddWarrantyLayoutSerial.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }

        binding.tiAddWarrantyEditDateStarting.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiAddWarrantyLayoutDateStarting.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiAddWarrantyLayoutDateStarting.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }

        binding.tiAddWarrantyEditDateExpiry.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiAddWarrantyLayoutDateExpiry.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiAddWarrantyLayoutDateExpiry.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }

        binding.tiAddWarrantyEditDescription.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiAddWarrantyLayoutDescription.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiAddWarrantyLayoutDescription.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }
    }

    private fun textInputsStrokeColor() {
        binding.tiAddWarrantyEditDateStarting.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(
                inputEmail: CharSequence?, start: Int, before: Int, count: Int
            ) {
                hideDateError()
            }

            override fun afterTextChanged(s: Editable?) {
            }
        })

        binding.tiAddWarrantyEditDateExpiry.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(
                inputEmail: CharSequence?, start: Int, before: Int, count: Int
            ) {
                hideDateError()
            }

            override fun afterTextChanged(s: Editable?) {
            }
        })
    }

    private fun categoryDropDown() {
        val titlesList = mutableListOf<String>()
        for (categoryTitleId in database.categoryDAO().allCategoryTitles) {
            titlesList.add(requireContext().getString(categoryTitleId))
        }

        binding.tiAddWarrantyDdCategory.setAdapter(
            ArrayAdapter(requireContext(), R.layout.dropdown_category, titlesList)
        )
    }

    private fun categoryDropDownSelection() {
        binding.tiAddWarrantyDdCategory.setOnItemClickListener { _, _, index, _ ->
            selectedCategory = database.categoryDAO().getCategoryById((index + 1).toString())

            selectedCategory?.let {
                binding.ivAddWarrantyIconCategory.setImageResource(it.icon)
            }
        }
    }

    private fun startingDatePicker() {
        binding.tiAddWarrantyEditDateStarting.inputType = InputType.TYPE_NULL
        binding.tiAddWarrantyEditDateStarting.keyListener = null

        binding.tiAddWarrantyEditDateStarting.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                openStartingDatePicker()
            }
        }
    }

    private fun expiryDatePicker() {
        binding.tiAddWarrantyEditDateExpiry.inputType = InputType.TYPE_NULL
        binding.tiAddWarrantyEditDateExpiry.keyListener = null

        binding.tiAddWarrantyEditDateExpiry.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                openExpiryDatePicker()
            }
        }
    }

    private fun openStartingDatePicker() {
        val startingDP = MaterialDatePicker.Builder.datePicker()
            .setTitleText(requireContext().getString(R.string.add_warranty_title_date_picker_starting))
            .setInputMode(MaterialDatePicker.INPUT_MODE_CALENDAR)
            .setSelection(MaterialDatePicker.todayInUtcMilliseconds())
            .build()

        startingDP.show(parentFragmentManager, Constants.FRAGMENT_TAG_ADD_CALENDAR_STARTING)

        startingDP.addOnPositiveButtonClickListener { selection ->
            startingCalendar = Calendar.getInstance()
            startingCalendar?.let {
                it.timeInMillis = selection
                startingDateInput = "${it.get(Calendar.YEAR)}-" +
                        "${decimalFormat.format((it.get(Calendar.MONTH)) + 1)}-" +
                        decimalFormat.format(it.get(Calendar.DAY_OF_MONTH))

                val startingDateInput =
                    "${decimalFormat.format((it.get(Calendar.MONTH)) + 1)}/" +
                            "${decimalFormat.format(it.get(Calendar.DAY_OF_MONTH))}/" +
                            "${it.get(Calendar.YEAR)}"

                binding.tiAddWarrantyEditDateStarting.setText(startingDateInput)
                binding.tiAddWarrantyEditDateStarting.clearFocus()
            }
        }

        startingDP.addOnDismissListener {
            binding.tiAddWarrantyEditDateStarting.clearFocus()
        }
    }

    private fun openExpiryDatePicker() {
        val expiryDP = MaterialDatePicker.Builder.datePicker()
            .setTitleText(requireContext().getString(R.string.add_warranty_title_date_picker_expiry))
            .setInputMode(MaterialDatePicker.INPUT_MODE_CALENDAR)
            .setSelection(MaterialDatePicker.todayInUtcMilliseconds())
            .build()

        expiryDP.show(parentFragmentManager, Constants.FRAGMENT_TAG_ADD_CALENDAR_EXPIRY)

        expiryDP.addOnPositiveButtonClickListener { selection ->
            expiryCalendar = Calendar.getInstance()
            expiryCalendar?.let {
                it.timeInMillis = selection
                expiryDateInput = "${it.get(Calendar.YEAR)}-" +
                        "${decimalFormat.format((it.get(Calendar.MONTH)) + 1)}-" +
                        decimalFormat.format(it.get(Calendar.DAY_OF_MONTH))

                val expiryDateInput =
                    "${decimalFormat.format((it.get(Calendar.MONTH)) + 1)}/" +
                            "${decimalFormat.format(it.get(Calendar.DAY_OF_MONTH))}/" +
                            "${it.get(Calendar.YEAR)}"

                binding.tiAddWarrantyEditDateExpiry.setText(expiryDateInput)
                binding.tiAddWarrantyEditDateExpiry.clearFocus()
            }
        }

        expiryDP.addOnDismissListener {
            binding.tiAddWarrantyEditDateExpiry.clearFocus()
        }
    }

    private fun returnToMainActivity() {
        binding.toolbarAddWarranty.setNavigationOnClickListener {
            requireActivity().onBackPressed()
        }
    }

    private fun addWarrantyOnClick() {
        binding.toolbarAddWarranty.menu[0]
            .setOnMenuItemClickListener {
                addWarranty()
                false
            }
    }

    private fun addWarranty() {
        val inputMethodManager = requireContext()
            .getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        inputMethodManager.hideSoftInputFromWindow(binding.root.applicationWindowToken, 0)

        if (NetworkHelper.hasNetworkAccess(requireContext())) {
            getWarrantyInput()
        } else {
            hideLoadingAnimation()
            Snackbar.make(
                binding.root,
                requireContext().getString(R.string.network_error_connection),
                LENGTH_INDEFINITE
            ).apply {
                setAction(requireContext().getString(R.string.network_error_retry)) { addWarranty() }
                show()
            }
        }
    }

    private fun getWarrantyInput() {
        val title = binding.tiAddWarrantyEditTitle.text.toString().trim()
        val brand = binding.tiAddWarrantyEditBrand.text?.toString()?.trim()
        val model = binding.tiAddWarrantyEditModel.text?.toString()?.trim()
        val serialNumber = binding.tiAddWarrantyEditSerial.text?.toString()?.trim()
        val description = binding.tiAddWarrantyEditDescription.text?.toString()?.trim()
        val categoryId = selectedCategory?.id ?: "10"

        if (title.isBlank()) {
            binding.tiAddWarrantyLayoutTitle.requestFocus()
            binding.tiAddWarrantyLayoutTitle.boxStrokeColor =
                ContextCompat.getColor(requireContext(), R.color.red)
        } else if (startingCalendar == null) {
            binding.tiAddWarrantyLayoutDateStarting.requestFocus()
            binding.tiAddWarrantyLayoutDateStarting.boxStrokeColor =
                ContextCompat.getColor(requireContext(), R.color.red)
        } else if (expiryCalendar == null) {
            binding.tiAddWarrantyLayoutDateExpiry.requestFocus()
            binding.tiAddWarrantyLayoutDateExpiry.boxStrokeColor =
                ContextCompat.getColor(requireContext(), R.color.red)
        } else if (!isStartingDateValid(startingCalendar!!, expiryCalendar!!)) {
            showDateError()
        } else {
            val warrantyInput = WarrantyInput(
                title, brand, model, serialNumber, startingDateInput,
                expiryDateInput, description, categoryId, Firebase.auth.currentUser?.uid.toString()
            )
            addWarrantyToFirestore(warrantyInput)
        }
    }

    private fun addWarrantyToFirestore(warrantyInput: WarrantyInput) {
        showLoadingAnimation()

        CoroutineScope(Dispatchers.IO).launch {
            try {
                warrantiesCollectionRef.add(warrantyInput).await()
                Log.i("addWarranty", "Warranty successfully added.")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    navController.navigate(R.id.action_addWarrantyFragment_to_warrantiesFragment)
                }
            } catch (e: Exception) {
                Log.e("addWarranty", "Exception: ${e.message}")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    Snackbar.make(
                        binding.root,
                        requireContext().getString(R.string.network_error_failure),
                        LENGTH_LONG
                    ).show()
                }
            }
        }
    }

    private fun showLoadingAnimation() {
        binding.toolbarAddWarranty.menu[0].isVisible = false
        binding.tiAddWarrantyEditTitle.isEnabled = false
        binding.tiAddWarrantyDdCategory.isEnabled = false
        binding.tiAddWarrantyEditBrand.isEnabled = false
        binding.tiAddWarrantyEditModel.isEnabled = false
        binding.tiAddWarrantyEditSerial.isEnabled = false
        binding.tiAddWarrantyEditDateStarting.isEnabled = false
        binding.tiAddWarrantyEditDateExpiry.isEnabled = false
        binding.tiAddWarrantyEditDescription.isEnabled = false
        binding.cpiAddWarranty.visibility = VISIBLE
    }

    private fun hideLoadingAnimation() {
        binding.toolbarAddWarranty.menu[0].isVisible = true
        binding.tiAddWarrantyEditTitle.isEnabled = true
        binding.tiAddWarrantyDdCategory.isEnabled = true
        binding.tiAddWarrantyEditBrand.isEnabled = true
        binding.tiAddWarrantyEditModel.isEnabled = true
        binding.tiAddWarrantyEditSerial.isEnabled = true
        binding.tiAddWarrantyEditDateStarting.isEnabled = true
        binding.tiAddWarrantyEditDateExpiry.isEnabled = true
        binding.tiAddWarrantyEditDescription.isEnabled = true
        binding.cpiAddWarranty.visibility = GONE
    }

    private fun showDateError() {
        binding.tiAddWarrantyLayoutDateStarting.requestFocus()
        binding.tiAddWarrantyLayoutDateStarting.boxStrokeColor =
            ContextCompat.getColor(requireContext(), R.color.red)
        binding.tvAddWarrantyDateError.visibility = VISIBLE
    }

    private fun hideDateError() {
        binding.tiAddWarrantyLayoutDateStarting.boxStrokeColor =
            ContextCompat.getColor(requireContext(), R.color.blue)
        binding.tvAddWarrantyDateError.visibility = GONE
    }

    private fun isStartingDateValid(startingCalendar: Calendar, expiryCalendar: Calendar): Boolean =
        expiryCalendar >= startingCalendar
}